moto/utilities/id_generator.py

```diff
@@ -0,0 +1,118 @@
+
+
+class ResourceIdentifier(abc.ABC):
+    """
+    Base class for resource identifiers. When implementing a new resource, it is important to set
+    the service and resource as they will be used to create the unique identifier for that resource.
+
+    It is recommended to implement the `generate` method using functions decorated with `@moto_id`.
+    This will ensure that your resource can be assigned a custom id.
+    """
+
+    service: str
+    resource: str
+
+    def __init__(self, account_id: str, region: str, name: str):
+        self.account_id = account_id
+        self.region = region
+        self.name = name or ""
+
+    @abc.abstractmethod
+    def generate(self, existing_ids: ExistingIds = None) -> str: ...
+
+    """ If `existing_ids` is provided, we will not return a custom id if it is already on the list"""
+
+    @property
+    def unique_identifier(self) -> str:
+        return ".".join(
+            [self.account_id, self.region, self.service, self.resource, self.name]
+        )
+
+
+class MotoIdManager:
+    """class to manage custom ids. Do not create instance and instead
+    use the `id_manager` instance created below."""
+
+    _custom_ids: Dict[str, str]
+    _id_sources: List[Callable[[ResourceIdentifier], Union[str, None]]]
+
+    _lock: threading.RLock
+
+    def __init__(self) -> None:
+        self._custom_ids = {}
+        self._lock = threading.RLock()
+        self._id_sources = []
+
+        self.add_id_source(self.get_custom_id)
+
+    def get_custom_id(
+        self, resource_identifier: ResourceIdentifier
+    ) -> Union[str, None]:
+        # retrieves a custom_id for a resource. Returns None
+        return self._custom_ids.get(resource_identifier.unique_identifier)
+
+    def set_custom_id(
+        self, resource_identifier: ResourceIdentifier, custom_id: str
+    ) -> None:
+        # Do not set a custom_id for a resource no value was found for the name
+        if not resource_identifier.name:
+            return
+        with self._lock:
+            self._custom_ids[resource_identifier.unique_identifier] = custom_id
+
+    def unset_custom_id(self, resource_identifier: ResourceIdentifier) -> None:
+        # removes a set custom_id for a resource
+        with self._lock:
+            self._custom_ids.pop(resource_identifier.unique_identifier, None)
+
+    def add_id_source(
+        self, id_source: Callable[[ResourceIdentifier], Union[str, None]]
+    ) -> None:
+        self._id_sources.append(id_source)
+
+    def find_id_from_sources(
+        self, resource_identifier: ResourceIdentifier, existing_ids: List[str]
+    ) -> Union[str, None]:
+        for id_source in self._id_sources:
+            if (
+                found_id := id_source(resource_identifier)
+            ) and found_id not in existing_ids:
+                return found_id
+        return None
+
+
+moto_id_manager = MotoIdManager()
+
+
+def moto_id(fn: Callable[..., str]) -> Callable[..., str]:
```

Comment:
Not really important in the bigger picture, but I guess this could become a `Callable[ParamSpec, str]`. 

That way we can remove the `type: ignore` on the `generate_str_id` method.

https://github.com/getmoto/moto/blob/5cbfacf3b1e4b214cf1c8f143a8b25ecd5213004/moto/core/responses.py#L224

Reply:
Actually, I tried that. I am not super experienced with mypy. But I was not able to get a scenario where a mix of explicit argument and `**kwargs` can be passed jointly. `Concatenate` did not help either.

I think the clarity and the simpler code in the decorator for using named parameters instead of being forced to use `*args, **kwargs` and deconstruct outweighs the loss on `type: ignore`

Reply:
Absolutely, that's fair - it is better readable like this